### PR TITLE
feat(core): route announcer

### PR DESCRIPTION
**Pull request by seyoon20087**
## Motivation

By doing this, client-side route changes will be announced to screen readers, improving accessibility to screen-reader users.

This was heavily inspired by https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/cache-dir/navigation.js#L173-L202.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

yes

## Test Plan

1. Add the route announcer component
2. Include the route announcer component in the root

## Related PRs

N/A

### CODE REVIEW

File: packages/docusaurus/src/client/theme-fallback/Root/index.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/theme-fallback/Root/index.tsx:22]: Please don't add it here. There's currently no way to swizzle root with CLI, so as soon as the user creates an empty `Root.js` file (which is also recommended in our docs), this gets lost. Maybe add it in `App.tsx`?

File: packages/docusaurus/package.json
Comment (Josh-Cena) [inline packages/docusaurus/package.json:51]: Do we necessarily want to use an external dep? We'd rather implement a simple portal ourselves, maybe even exposing it as a core client API (as we did for `ErrorBoundary`), so that (a) we don't get blocked on an external dep (b) we can control bundle size and remove code that we don't need. Can we make our own abstraction over `React.createPortal`?

File: packages/docusaurus/src/client/routeAnnouncer.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/routeAnnouncer.tsx:41]: This timeout needs to be cleaned up. Also, we can simply do `document.title ?? content ?? pathname` I believe?

File: packages/docusaurus/src/client/routeAnnouncer.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/routeAnnouncer.tsx:45]: I'm not sure query changes should be announced. Docusaurus builds a static site and query should not lead to significant page changes.

**Review by Josh-Cena — COMMENTED** [review]
Thanks, seems to work great 👍 Will leave it to others to review as well. Left some more suggestions.

File: packages/docusaurus/src/client/routeAnnouncer.tsx
Comment (seyoon20087) [reply packages/docusaurus/src/client/routeAnnouncer.tsx]: I have removed the line

File: packages/docusaurus/src/client/routeAnnouncer.tsx
Comment (seyoon20087) [reply packages/docusaurus/src/client/routeAnnouncer.tsx]: I did a commit for `document.title ?? content ?? pathname`. But, unfortunately when I don't include the timeout the announcer will keep the previous content (in my testing).

Comment (Josh-Cena) [reply packages/docusaurus/src/client/routeAnnouncer.tsx]: That's expected, because our route components are lazy-loaded with react-loadable, so when the router calls `ComponentCreator`, the new component hasn't mounted yet. We may have to do some refactors to that as I'm not really sure this timeout will work in all cases (e.g. if rendering is blocked)

File: packages/docusaurus/package.json
Comment (seyoon20087) [reply packages/docusaurus/package.json]: I have updated it. Please have a look

File: packages/docusaurus/src/client/exports/Portal.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/exports/Portal.tsx:20]:
```suggestion
}: PortalProps): React.ReactPortal | null {
```

File: yarn.lock
Comment (Josh-Cena) [inline yarn.lock:3214]: Re-run install to clear this

**Comment by Josh-Cena** [comment]
The timeout unfortunately doesn't work if the network is slow and the bundle preloading takes more than 50ms. I think this should better be implemented as a client module since it seems pretty decoupled from the main app. We can do this after fixing #6732

File: packages/docusaurus/src/client/App.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/App.tsx:28]: Please please please, ESLint and TS exist for a reason... You can't just disable them both...

File: packages/docusaurus/src/client/RouteAnnouncerWrapper.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/RouteAnnouncerWrapper.tsx:27]: Just to confirm: the reason you are using `componentDidUpdate` (and hence a class component) is on purpose, because you want the announcement to happen synchronously, correct?

Comment (seyoon20087) [reply packages/docusaurus/src/client/RouteAnnouncerWrapper.tsx:37]: I am using this because the route announcer can work correctly, even on slow networks. I tested on my side, and it works OK (like Gatsby 4 did)

Comment (seyoon20087) [reply packages/docusaurus/src/client/RouteAnnouncerWrapper.tsx:37]: I'm also using this lifecycle because it doesn't trigger on the initial pre-render

Comment (Josh-Cena) [reply packages/docusaurus/src/client/RouteAnnouncerWrapper.tsx:37]: `useEffect` would work just the same, from some quick eyeballing? Have you tried?

Comment (Josh-Cena) [reply packages/docusaurus/src/client/RouteAnnouncerWrapper.tsx:37]: > doesn't trigger on the initial pre-render

pre-rendering, or hydration? Even if you want to skip hydration, you can use a `mounted` ref to bypass this. We only fall back to class component if function component can't work at all (e.g. `componentDidCatch`).

Comment (Josh-Cena) [reply packages/docusaurus/src/client/RouteAnnouncerWrapper.tsx]: Actually, scratch it. I'm fine with what we have now; `componentDidUpdate` seems like a good fit.

File: packages/docusaurus/src/client/RouteAnnouncer.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/RouteAnnouncer.tsx:38]: No need to fallback. `document.title` always exists because we have it in core `SiteMetadataDefaults`.

**Review by Josh-Cena — APPROVED** [review]
The current behavior looks cool to me. I still somehow believe this can be factored out into a separate client module and not in the main component tree (since all it does is consuming the location and portaling to a DOM element), but I'll let @slorber take a look first. At least now we can be unblocked from #6732

**Review by lex111 — COMMENTED** [review]
Perhaps we should move this component to the classic-theme area to make related message translatable?

**Comment by Josh-Cena** [comment]
I agree. Given the current architecture, it seems feasible to put this in layout as well.

File: packages/docusaurus/src/client/RouteAnnouncer.tsx
Comment (slorber) [inline packages/docusaurus/src/client/RouteAnnouncer.tsx:36]: accessibility label should be translatable => in theme

File: packages/docusaurus/src/client/RouteAnnouncer.tsx
Comment (slorber) [inline packages/docusaurus/src/client/RouteAnnouncer.tsx:34]: this is a theme-related selector, so it should rather be in the theme

(although it's unlikely 2 themes will have different selectors 🤷‍♂️ )

File: packages/docusaurus/src/client/exports/Portal.tsx
Comment (slorber) [inline packages/docusaurus/src/client/exports/Portal.tsx:18]: are other portal types planned? 

What is the purpose of calling `document.createElement("docusaurus-portal");` ?

File: packages/docusaurus/src/client/RouteAnnouncer.tsx
Comment (slorber) [inline packages/docusaurus/src/client/RouteAnnouncer.tsx:52]: what's the purpose of this other portal type?

and why do we need a portal in the first place?

(I don't know how route announcement is supposed to work)

File: packages/docusaurus/src/client/RouteAnnouncer.tsx
Comment (slorber) [inline packages/docusaurus/src/client/RouteAnnouncer.tsx:51]: does this component need a children? 

Can't it work standalone like `<RouteAnnouncer/>` ?

Afaik it doesn't provide any context to the children so that props seem unnecessary

File: packages/docusaurus/src/client/RouteAnnouncer.tsx
Comment (slorber) [inline packages/docusaurus/src/client/RouteAnnouncer.tsx:22]: does it need to be a class component?

File: packages/docusaurus/src/client/RouteAnnouncer.tsx
Comment (slorber) [inline packages/docusaurus/src/client/RouteAnnouncer.tsx:33]: why is requestAnimationFrame needed?

**Review by slorber — CHANGES_REQUESTED** [review]
Hey, looks like an interesting addition.

I don't know how route announcement is supposed to be implemented theoretically, can you explain, please? Particularly the thing about using a custom element, a portal, rAF...

I don't understand a few choices in this PR that looks weird to me, but maybe I'm wrong

Maybe this feature could be a route announcement plugin using a client module?

The announcement label template should rather be translatable

File: packages/docusaurus/src/client/exports/Portal.tsx
Comment (Josh-Cena) [reply packages/docusaurus/src/client/exports/Portal.tsx:18]: @slorber This is the container node's tag name. Portal elements can often be custom elements so as not to have any default user-agent styles.

File: packages/docusaurus-theme-classic/src/routeAnnouncer.css
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/routeAnnouncer.css:8]: Using an ID selector here. Not sure if we necessarily want a class name here, or if we want to use inline styles.

File: packages/docusaurus-theme-classic/src/routeAnnouncer.ts
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/routeAnnouncer.ts:19]: Clear the announcement from previous route transition. Technically, we can put it in the cleanup function of `onRouteDidUpdate`, but we don't have a clear semantic of what that cleanup should do.

File: packages/docusaurus-theme-classic/src/routeAnnouncer.ts
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/routeAnnouncer.ts:28]: Pure DOM manipulation. Because client modules don't have access to the React context, it would be quite cumbersome if we want to use React to do this. The alternative will be to put this in `Layout`, but there may be other edge-cases (for example, the layout doesn't re-mount between doc pages)

File: website/testCSSOrder.mjs
Comment (Josh-Cena) [inline website/testCSSOrder.mjs:51]: I'm surprised that this moved, but it makes more sense this way? Not sure why it was loaded so early before.

**Review by Josh-Cena — COMMENTED** [review]
@slorber @lex111 I've move the announcer implementation to theme-classic so we can translate it. Because the client module has just been fixed, I decided to go with a pure-DOM solution without wiring it to a theme component, so that we have a use-case of reading and manipulating DOM in client module. I doubt if anyone wants to swizzle the announcement anyways. I've kept the `Portal` core component for now, although we don't use it anymore, in case it's useful in the future. Tell me what you think.

File: packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx
Comment (Josh-Cena) [inline packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx:52]: Exposed the cleanup function here, so we can clean up the side-effect of appending DOM child.

**Pull request closed without merging**
